# Data Exploration Notebook

_For USD-599 Capstone Project by Hunter Blum, Kyle Esteban Dalope, and Nicholas Lee (Summer 2023)_

In [1]:
# Library Imports
import pandas as pd

In [2]:
# Read in listings tables
june2022_listings_df = pd.read_csv("../Data/2022_06_june_listings.csv.gz", compression = "gzip")
sept2022_listings_df = pd.read_csv("../Data/2022_09_sept_listings.csv.gz", compression = "gzip")
dec2022_listings_df = pd.read_csv("../Data/2022_12_dec_listings.csv.gz", compression = "gzip")
mar2023_listings_df = pd.read_csv("../Data/2023_03_mar_listings.csv.gz", compression = "gzip")

# Sample Table
june2022_listings_df.head(5)

id                                      listing_url   
0            53661776            https://www.airbnb.com/rooms/53661776  \
1  599610579640991142  https://www.airbnb.com/rooms/599610579640991142   
2            52088878            https://www.airbnb.com/rooms/52088878   
3  628640985640584632  https://www.airbnb.com/rooms/628640985640584632   
4            19621314            https://www.airbnb.com/rooms/19621314   

        scrape_id last_scraped   
0  20220615002624   2022-06-15  \
1  20220615002624   2022-06-15   
2  20220615002624   2022-06-15   
3  20220615002624   2022-06-15   
4  20220615002624   2022-06-15   

                                                name   
0   2 bedrooms/ 1 bath - Upstairs - Near Lake Hodges  \
1  RelaxingSunset View PrivateBalcony for SIX guests   
2          Cozy independent room next to Lake Murray   
3  Bright freshly renovated apartment w/ kitchenette   
4    Huge Royal Luxury En Suite (Bath, WC & Balcony)   

                                         description   
0  This listing consists of 2 bedrooms and 1 bath...  \
1  Stylishly appointed, 1,313 Square foot sunset ...   
2  Spacious private room in a quiet residential a...   
3  This unique place has a style all its own read...   
4  Royal Feeling-Resort style, private gate, Soot...   

                               neighborhood_overview   
0       The neighborhood is very quiet and peaceful.  \
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Very quiet neighbors and country side living y...   

                                         picture_url    host_id   
0  https://a0.muscache.com/pictures/miso/Hosting-...   62320777  \
1  https://a0.muscache.com/pictures/miso/Hosting-...  414789390   
2  https://a0.muscache.com/pictures/miso/Hosting-...  100368579   
3  https://a0.muscache.com/pictures/miso/Hosting-...   95926933   
4  https://a0.muscache.com/pictures/e1811785-3d44...   46170232   

                                      host_url  ...   
0   https://www.airbnb.com/users/show/62320777  ...  \
1  https://www.airbnb.com/users/show/414789390  ...   
2  https://www.airbnb.com/users/show/100368579  ...   
3   https://www.airbnb.com/users/show/95926933  ...   
4   https://www.airbnb.com/users/show/46170232  ...   

  review_scores_communication review_scores_location review_scores_value   
0                         NaN                    NaN                 NaN  \
1                        5.00                   5.00                5.00   
2                        4.86                   4.86                4.71   
3                         NaN                    NaN                 NaN   
4                        4.99                   4.93                5.00   

  license instant_bookable calculated_host_listings_count   
0     NaN                f                              7  \
1     NaN                t                              2   
2     NaN                f                              1   
3     NaN                t                             11   
4     NaN                f                              5   

  calculated_host_listings_count_entire_homes   
0                                           2  \
1                                           2   
2                                           0   
3                                          11   
4                                           0   

  calculated_host_listings_count_private_rooms   
0                                            5  \
1                                            0   
2                                            1   
3                                            0   
4                                            5   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0               NaN  
1                                           0              0.64  
2     

## Table Features
Exploration of table shapes and features

In [3]:
# Individual Table Dimensions
def table_dim(df_shape):
    print("There are {} rows and {} columns".format(df_shape[0], df_shape[1]))

print("Table Dimensions for June 2022 Listings: ")
table_dim(june2022_listings_df.shape)
print("\nTable Dimensions for Sept. 2022 Listings: ")
table_dim(sept2022_listings_df.shape)
print("\nTable Dimensions for Dec. 2022 Listings: ")
table_dim(dec2022_listings_df.shape)
print("\nTable Dimensions for Mar. 2022 Listings: ")
table_dim(mar2023_listings_df.shape)

Table Dimensions for June 2022 Listings: 
There are 13272 rows and 74 columns

Table Dimensions for Sept. 2022 Listings: 
There are 14188 rows and 75 columns

Table Dimensions for Dec. 2022 Listings: 
There are 12781 rows and 75 columns

Table Dimensions for Mar. 2022 Listings: 
There are 12871 rows and 75 columns


The above inspection of listing dimensions reveals that after June 2022, a new, 75th column was added to the data collected.

In [4]:
# Get column names from each table
june_col = june2022_listings_df.columns
sept_col = sept2022_listings_df.columns
dec_col = dec2022_listings_df.columns
mar_col = mar2023_listings_df.columns


print(sept_col.difference(june_col))
print(sept_col.difference(dec_col))
print(sept_col.difference(mar_col))

Index(['source'], dtype='object')
Index([], dtype='object')
Index([], dtype='object')


The new column added after June 2022, was called "source."

In [5]:
sept2022_listings_df["source"].describe()

count           14188
unique              2
top       city scrape
freq            12956
Name: source, dtype: object

In [6]:
dec2022_listings_df["source"].describe()

count           12781
unique              2
top       city scrape
freq            11589
Name: source, dtype: object

In [7]:
mar2023_listings_df["source"].describe()

count           12871
unique              2
top       city scrape
freq            11644
Name: source, dtype: object

Within the _source_ column, there are two unique values, where one value dominates the majority of records (e.g. in records for the quarter up until March 2023, the predominant source value occupied 11,644 of 12,871 records). In planning for future machine learning applications, such a column would not be greatly informative. Therefore, it can be removed, and the tables may be joined along the unique house id.

### Dropping "source" column

In [8]:
# Removing the source column
sept2022_listings_df = sept2022_listings_df.drop(columns="source", axis = 1)
dec2022_listings_df = dec2022_listings_df.drop(columns = "source", axis = 1)
mar2023_listings_df = mar2023_listings_df.drop(columns = "source", axis = 1)

## Combining all tables into one dataframe
The four quarterly tables are concatenated to form one dataframe. Then, duplicates from the _id_ column are removed, keeping the last, most recent, up-to-date entry for each listing.

In [18]:
# Combine all four dataframes into one
all_data = pd.concat([june2022_listings_df, sept2022_listings_df,
                        dec2022_listings_df, mar2023_listings_df])

In [19]:
# Convert last_scraped column to datetime data type
all_data["last_scraped"] = pd.to_datetime(all_data["last_scraped"])

In [20]:
# Sort by date and drop duplicates (keeping most recent entries)
all_data_no_dupes = all_data.sort_values(
    "last_scraped").drop_duplicates(["id"], keep = "last")

## Merged Dataframe Descriptive Statistics

In [22]:
# Check new dataframe dimensions
all_data_no_dupes.shape

(18627, 74)

In [23]:
all_data_no_dupes.describe()

id     scrape_id                   last_scraped   
count  1.862700e+04  1.862700e+04                          18627  \
mean   2.449705e+17  2.022742e+13  2023-01-27 14:34:06.706393856   
min    6.000000e+00  2.022062e+13            2022-06-15 00:00:00   
25%    2.994373e+07  2.022122e+13            2022-12-24 00:00:00   
50%    4.981844e+07  2.023032e+13            2023-03-25 00:00:00   
75%    6.287697e+17  2.023032e+13            2023-03-25 00:00:00   
max    8.536537e+17  2.023032e+13            2023-03-25 00:00:00   
std    3.334764e+17  4.340641e+09                            NaN   

            host_id  host_listings_count  host_total_listings_count   
count  1.862700e+04         18576.000000               18576.000000  \
mean   1.555625e+08            84.257321                 121.971307   
min    2.900000e+01             0.000000                   0.000000   
25%    2.718982e+07             1.000000                   2.000000   
50%    9.578050e+07             4.000000                   5.000000   
75%    2.625391e+08            20.000000                  29.000000   
max    5.063270e+08          4807.000000                7816.000000   
std    1.546986e+08           374.064936                 525.947790   

       neighbourhood_group_cleansed      latitude     longitude  accommodates   
count                           0.0  18627.000000  18627.000000  18627.000000  \
mean                            NaN     32.763184   -117.172498      4.815483   
min                             NaN     32.519860   -117.282580      0.000000   
25%                             NaN     32.718332   -117.246039      2.000000   
50%                             NaN     32.754850   -117.164620      4.000000   
75%                             NaN     32.798349   -117.129225      6.000000   
max                             NaN     33.125690   -116.896560     16.000000   
std                             NaN      0.092287      0.073437      3.168323   

       ...  review_scores_cleanliness  review_scores_checkin   
count  ...               15415.000000           15413.000000  \
mean   ...                   4.745367               4.842457   
min    ...                   1.000000               1.000000   
25%    ...                   4.670000               4.840000   
50%    ...                   4.860000               4.950000   
75%    ...                   4.990000               5.000000   
max    ...                   5.000000               5.000000   
std    ...                   0.391430               0.347829   

       review_scores_communication  review_scores_location   
count                 15415.000000            15413.000000  \
mean                      4.831073                4.814729   
min                       1.000000                1.000000   
25%                       4.820000                4.780000   
50%                       4.950000                4.910000   
75%                       5.000000                5.000000   
max                       5.000000                5.000000   
std                       0.364334                0.323353   

       review_scores_value  calculated_host_listings_count   
count          15413.00000                    18627.000000  \
mean               4.67671                       15.837709   
min                1.00000                        1.000000   
25%                4.59000                        1.000000   
50%                4.78000                        3.000000   
75%                4.91000                       14.000000   
max                5.00000                      216.000000   
std                0.40645                       30.835294   

       calculated_host_listings_count_entire_homes   
count                                 18627.000000  \
mean                                     14.491813   
min                                       0.000000   
25%                                       1.000000   
50%                                       2.000000   
75%            